### PR TITLE
Add more aggressive and fine grained throttling logic

### PR DESCRIPTION
**Pull request by larskuhtz**

### CODE REVIEW

File: src/Chainweb/Chainweb.hs [hunk 1]
```diff
@@ -278,6 +305,7 @@ instance FromJSON (ChainwebConfiguration -> ChainwebConfiguration) where
         <*< configTransactionIndex %.: "transactionIndex" % o
         <*< configIncludeOrigin ..: "includeOrigin" % o
         <*< configThrottleRate ..: "throttleRate" % o
+        <*< configMiningThrottleRate ..: "miningThrottleRate" % o
```
Comment (larskuhtz) [inline src/Chainweb/Chainweb.hs]:
```suggestion
        <*< configMiningThrottleRate ..: "miningThrottleRate" % o
        <*< configPutPeerThrottleRate ..: "putPeerThrottleRate" % o
```

File: src/Chainweb/Chainweb.hs [hunk 2]
```diff
@@ -241,7 +264,9 @@ defaultChainwebConfiguration v = ChainwebConfiguration
     , _configP2p = defaultP2pConfiguration
     , _configTransactionIndex = defaultEnableConfig defaultTransactionIndexConfig
     , _configIncludeOrigin = True
-    , _configThrottleRate = 1000
+    , _configThrottleRate = 25 -- 25 per second
+    , _configMiningThrottleRate = 1 -- 1 per second
+    , _configPutPeerThrottleRate = 1 -- one per second
```
Comment (larskuhtz) [inline src/Chainweb/Chainweb.hs]:
```suggestion
    , _configPutPeerThrottleRate = 11 -- 11 per second, one for each p2p network
```

File: src/Chainweb/Chainweb.hs [hunk 3]
```diff
@@ -241,7 +264,9 @@ defaultChainwebConfiguration v = ChainwebConfiguration
     , _configP2p = defaultP2pConfiguration
     , _configTransactionIndex = defaultEnableConfig defaultTransactionIndexConfig
     , _configIncludeOrigin = True
-    , _configThrottleRate = 1000
+    , _configThrottleRate = 25 -- 25 per second
+    , _configMiningThrottleRate = 1 -- 1 per second
```
Comment (larskuhtz) [inline src/Chainweb/Chainweb.hs]:
```suggestion
    , _configMiningThrottleRate = 2 --  per second
```

File: src/Chainweb/Chainweb.hs [hunk 4]
```diff
@@ -241,7 +264,9 @@ defaultChainwebConfiguration v = ChainwebConfiguration
     , _configP2p = defaultP2pConfiguration
     , _configTransactionIndex = defaultEnableConfig defaultTransactionIndexConfig
     , _configIncludeOrigin = True
-    , _configThrottleRate = 1000
+    , _configThrottleRate = 25 -- 25 per second
```
Comment (larskuhtz) [inline src/Chainweb/Chainweb.hs]:
```suggestion
    , _configThrottleRate = 50 -- per second
```

File: src/Chainweb/Chainweb.hs [hunk 5]
```diff
@@ -618,6 +669,51 @@ withChainwebInternal conf logger peer rocksDb dbDir nodeid resetDb inner = do
...
+mkMiningThrottler rate = mkThrottler 5 rate (checkPathPrefix ["mining", "work"])
+
+mkPutPeerThrottler :: Double -> IO (Throttle Address)
+mkPutPeerThrottler rate = mkThrottler 5 rate $ \r ->
+    elem "peer" (pathInfo r) && requestMethod r == "PUT"
+
+mkLocalThrottler :: Double -> IO (Throttle Address)
+mkLocalThrottler rate = mkThrottler 5 rate (checkPathPrefix path)
+  where
+    path = ["pact", "api", "v1", "local"]
```
Comment (fosskers) [inline src/Chainweb/Chainweb.hs]: Correct path? @emmanueldenloye @emilypi ?

File: src/P2P/Node/PeerDB.hs [hunk 6]
```diff
@@ -287,7 +298,9 @@ addPeerInfo nid = addPeerEntry . newPeerEntry nid
 -- is delete for all network ids.
 --
 deletePeer :: PeerInfo -> PeerSet -> PeerSet
-deletePeer i = deleteIx (_peerAddr i)
+deletePeer i s = case _peerEntrySticky <$> (getOne $ getEQ (_peerAddr i) s) of
+    Just True -> s
+    _ -> deleteIx (_peerAddr i) s
```
Comment (larskuhtz) [inline src/P2P/Node/PeerDB.hs]: Don't delete sticky nodes

File: src/P2P/Node/PeerDB.hs [hunk 7]
```diff
@@ -357,6 +370,9 @@ peerDbInsertList peers (PeerDb lock var) =
 peerDbInsertPeerInfoList :: NetworkId -> [PeerInfo] -> PeerDb -> IO ()
 peerDbInsertPeerInfoList nid ps = peerDbInsertList (newPeerEntry nid <$> ps)
```
Comment (larskuhtz) [inline src/P2P/Node/PeerDB.hs]: This default to inserting non-sticky nodes

File: src/P2P/Node/RestAPI/Server.hs [hunk 8]
```diff
@@ -95,6 +95,7 @@ peerGetHandler db nid limit next = do
         . SP.zip (SP.each [0..])
         . SP.each
         . toAscList (Proxy @HostAddressIdx)
+        . getEQ (SuccessiveFailures 0)
```
Comment (larskuhtz) [inline src/P2P/Node/RestAPI/Server.hs]: Only gossip around nodes that didn't have any recent failure.

(Is this too strict? Should we allow nodes with, say, one successive failure?)

File: src/Chainweb/Chainweb/MinerResources.hs [hunk 9]
```diff
@@ -86,14 +87,15 @@ withMiningCoordination logger enabled cutDb inner
             { _coordLogger = logger
             , _coordCutDb = cutDb
             , _coordState = t
-            , _coordLimit = 2500
+            , _coordLimit = 1200
```
Comment (larskuhtz) [inline src/Chainweb/Chainweb/MinerResources.hs]: We throttle with 503 if we have more than 1200 pending work items.

Should we make this configurable?

File: src/P2P/Node.hs [hunk 10]
```diff
@@ -502,10 +502,10 @@ findNextPeer conf node = do
...
     -- TODO: how expensive is this? should be cache the classification?
     --
-    let p0 = IXS.getGT (ActiveSessionCount 0) $ IXS.getEQ (SuccessiveFailures 0) base
-        p1 = IXS.getEQ (ActiveSessionCount 0) $ IXS.getEQ (SuccessiveFailures 0) base
-        p2 = IXS.getGT (ActiveSessionCount 0) $ IXS.getGTE (SuccessiveFailures 1) base
-        p3 = IXS.getEQ (ActiveSessionCount 0) $ IXS.getGTE (SuccessiveFailures 1) base
+    let p0 = IXS.getGT (ActiveSessionCount 0) $ IXS.getLTE (SuccessiveFailures 1) base
+        p1 = IXS.getEQ (ActiveSessionCount 0) $ IXS.getLTE (SuccessiveFailures 1) base
+        p2 = IXS.getGT (ActiveSessionCount 0) $ IXS.getGT (SuccessiveFailures 1) base
+        p3 = IXS.getEQ (ActiveSessionCount 0) $ IXS.getGT (SuccessiveFailures 1) base
```
Comment (larskuhtz) [inline src/P2P/Node.hs]: I wonder, if we need to add policy for sticky nodes here. Although, doing so may increase the likelihood of forks. So, let's be careful and don't do it for now.

But this needs overhaul in any case.

File: src/Chainweb/Chainweb/MinerResources.hs (hunk 9 above)
Comment (fosskers) [reply src/Chainweb/Chainweb/MinerResources.hs]: Yes.

File: src/P2P/Node/RestAPI/Server.hs (hunk 8 above)
Comment (fosskers) [reply src/P2P/Node/RestAPI/Server.hs]: Perhaps. @gregorycollins ?

File: src/Chainweb/Chainweb.hs (hunk 5 above)
Comment (emilypi) [reply src/Chainweb/Chainweb.hs]: correct

**Pull request merged**